A40-csds-support.md

```diff
@@ -0,0 +1,271 @@
+  // server. The client will not fail requests, but will queue them until update
+  // arrives or the client times out waiting for the resource.
+  REQUESTED = 1;
+
+  // This resource has been requested by the client but has either not been
+  // delivered by the server or was previously delivered by the server and then
+  // subsequently removed from resources provided by the server. For more
+  // information, please refer to the :ref:`"Knowing When a Requested Resource
+  // Does Not Exist" <xds_protocol_resource_not_existed>` section.
+  DOES_NOT_EXIST = 2;
+
+  // Client received this resource and replied with ACK.
+  ACKED = 3;
+
+  // Client received this resource and replied with NACK.
+  NACKED = 4;
+}
+```
+
+### ADS Parsing Logic Update: Continue After First Error
+
+Current gRPC ADS parsing logic is when the response parser observes an error in
+the ADS response, it fails the entire message and aborts parsing.
+
+To improve the debuggability of gRPC, gRPC needs to populate **the reason** and
+**affected resources** when rejecting an update response. This information
+requires the ADS parser to continue parsing past the first error. Here are the
+expected behavior under scenarios:
+
+* If the entire message won’t parse: no need to record anything in CSDS, since
+  we don’t know what type of xDS config or what resources are being updated;
+* If one resource won’t parse: don’t abort the parsing, record the error in the
+  error string. The error string will be attached to other resources that are
+  successfully recognized;
+* If one resource has validation error: record the error, and attach the error
+  string to all resources including itself.
+
+Here is an example of NACK information handling with new ADS changes:
+
+```
+# Imagine we have endpoint A, B, C
+EDS -> {A, B, C}, version 1
+gRPC -> ACK
+CSDS -> [
+  {Endpoint A, ACK, version 1},
+  {Endpoint B, ACK, version 1},
+  {Endpoint C, ACK, version 1}
+]
+
+# The newer endpoint B contains parsing error
+EDS -> {A, B}, version 2
+gRPC -> NACK, Failed to parse endpoint B 
+CSDS -> [
+  {Endpoint A, NACK, version 1, rejected version 2, rejected reason: Failed to parse endpoint B},
+  {Endpoint B, NACK, version 1, rejected version 2, rejected reason: Failed to parse endpoint B},
+  {Endpoint C, ACK, version 1}
+]
+
+# Accepted update will clean error states
+EDS -> {B, C} version 3
+gRPC -> ACK
+CSDS -> [
+  {Endpoint A, NACK, version 1, rejected version 2, rejected reason: Failed to parse endpoint B},
+  {Endpoint B, ACK, version 3},
+  {Endpoint C, ACK, version 3}
+]
+```
+
+
+### xDS Config Error State
+
+When an ADS response is rejected, gRPC should provide debug information via
+CSDS. This is done via `UpdateFailureState` within the `config_dump.proto`,
+which includes the version, timestamp, and the reason of the rejected update
+(see [proto
+definition](https://github.com/envoyproxy/envoy/blob/main/api/envoy/admin/v3/config_dump.proto#L72)):
```

Comment:
(suggestion) Line numbers tend to change, consider replacing with a permalink.

Reply:
Link updated.